context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java

```diff
@@ -0,0 +1,287 @@
+          return i;
+        }
+      }
+      return -1;
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder valuesSb = new StringBuilder();
+      valuesSb.append("CollisionLeaf(");
+      for (int i = 0; i < values.length; i++) {
+        valuesSb.append("(key=").append(keys[i]).append(" value=").append(values[i]).append(") ");
+      }
+      return valuesSb.append(")").toString();
+    }
+  }
+
+  // Not actually annotated to avoid depending on guava
+  // @VisibleForTesting
+  static final class CompressedIndex<K,V> implements Node<K,V> {
+    private static final int BITS = 5;
+    private static final int BITS_MASK = 0x1F;
+
+    final int bitmap;
+    final Node<K,V>[] values;
+
+    private CompressedIndex(int bitmap, Node<K,V>[] values) {
+      this.bitmap = bitmap;
+      this.values = values;
+    }
+
+    @Override
+    public V get(K key, int hash, int bitsConsumed) {
+      int indexBit = indexBit(hash, bitsConsumed);
+      if ((bitmap & indexBit) == 0) {
+        return null;
+      }
+      int compressedIndex = compressedIndex(indexBit);
+      return values[compressedIndex].get(key, hash, bitsConsumed + BITS);
+    }
+
+    @Override
+    public Node<K,V> put(K key, V value, int hash, int bitsConsumed) {
+      int indexBit = indexBit(hash, bitsConsumed);
+      int compressedIndex = compressedIndex(indexBit);
+      if ((bitmap & indexBit) == 0) {
+        // Insert
+        int newBitmap = bitmap | indexBit;
+        @SuppressWarnings("unchecked")
+        Node<K,V>[] newValues = (Node<K,V>[]) new Node<?,?>[values.length + 1];
+        System.arraycopy(values, 0, newValues, 0, compressedIndex);
+        newValues[compressedIndex] = new Leaf<K,V>(key, value);
+        System.arraycopy(
+            values,
+            compressedIndex,
+            newValues,
+            compressedIndex + 1,
+            values.length - compressedIndex);
+        return new CompressedIndex<K,V>(newBitmap, newValues);
+      } else {
+        // Replace
+        Node<K,V>[] newValues = Arrays.copyOf(values, values.length);
+        newValues[compressedIndex] =
+            values[compressedIndex].put(key, value, hash, bitsConsumed + BITS);
+        return new CompressedIndex<K,V>(bitmap, newValues);
+      }
+    }
+
+    static <K,V> Node<K,V> combine(
+        Node<K,V> node1, int hash1, Node<K,V> node2, int hash2, int bitsConsumed) {
+      assert hash1 != hash2;
+      int indexBit1 = indexBit(hash1, bitsConsumed);
+      int indexBit2 = indexBit(hash2, bitsConsumed);
+      if (indexBit1 == indexBit2) {
+        Node<K,V> node = combine(node1, hash1, node2, hash2, bitsConsumed + BITS);
+        @SuppressWarnings("unchecked")
+        Node<K,V>[] values = (Node<K,V>[]) new Node<?,?>[] {node};
+        return new CompressedIndex<K,V>(indexBit1, values);
+      } else {
+        // Make node1 the smallest
+        if (uncompressedIndex(hash1, bitsConsumed) > uncompressedIndex(hash2, bitsConsumed)) {
```

Comment:
If `indexBit1` is (1<<31) then we would mistakenly think it's a negative number of a large magnitude. I think we'd need to do some bit twiddling like `(indexBit1 - 1) & indexBit2 != 0` but it would be less readable.